### PR TITLE
Add tls record fragmentation

### PR DESCRIPTION
**Pull request by myleshorton**

### CODE REVIEW

File: chained/proxy.go [hunk 1]
```diff
@@ -324,36 +323,6 @@ func newProxy(name, addr, protocol, network string, s *config.ProxyConfig, uc co
...
-			dfConn, err := p.reportedDial(func(op *ops.Op) (net.Conn, error) { return dial(network, addr) })
-			dfConn, err = overheadWrapper(true)(dfConn, op.FailIf(err))
-			if err == nil {
-				dfConn = idletiming.Conn(dfConn, IdleTimeout, func() {
-					log.Debug("enhttp connection idled")
-				})
-			}
-			return dfConn, err
-		}
-	}
```
Comment (myleshorton) [inline chained/proxy.go:356]: We don't use this anymore

File: chained/proxy.go [hunk 2]
```diff
@@ -114,6 +110,9 @@ func CreateDialer(configDir, name string, s *config.ProxyConfig, uc common.UserC
 		return nil, err
 	}
 	p, err := newProxy(name, addr, transport, network, s, uc)
+	if err != nil {
+		return nil, err
+	}
```
Comment (myleshorton) [inline chained/proxy.go:117]: drive-by

File: chained/https_impl.go [hunk 3]
```diff
@@ -104,7 +109,7 @@ func (impl *httpsImpl) dialServer(op *ops.Op, ctx context.Context) (net.Conn, er
 func timeoutFor(ctx context.Context) time.Duration {
 	deadline, ok := ctx.Deadline()
 	if ok {
-		return deadline.Sub(time.Now())
+		return time.Until(deadline)
```
Comment (myleshorton) [inline chained/https_impl.go:117]: drive-by

File: chained/proxy.go (hunk 2 above)
Comment (hwh33) [reply chained/proxy.go:117]: Wow, good catch!

File: chained/tlsfrag.go [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+		if err != nil {
+			return conn
+		}
+		return tlsFragConn
+	}
+	return conn
+}
+
+func splitHelloInfo(pc *config.ProxyConfig) (bool, int) {
+	indexStr, ok := pc.PluggableTransportSettings["tlsfrag_split_index"]
```
Comment (hwh33) [inline chained/tlsfrag.go]: Could you implement this in a way that would support more complex future configurations?  The two I currently have in mind are (1) random within a range and (2) splitting the SNI.

My suggestion would be to rename the option `tlsfrag` and expect a value of the form `frag mechanism, { mechanism-specific config }`.  Example values would be `index:7`, `sni` (randomly within the SNI), `sni:3` (byte 3 of the SNI), `random` (randomly anywhere), `random:5,100` (random within bytes 5 to 100).

As-is, it would be a little awkward to add one of these alternative fragmentation mechanisms in a backwards-compatible way.  For example, if we add a new option `tlsfrag_sni`, what does the client do when both that and `tlsfrag_split_index` are specified?

File: chained/tlsfrag.go [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+func tlsFragConn(conn net.Conn, pc *config.ProxyConfig) net.Conn {
+	splitHello, index := splitHelloInfo(pc)
+	if splitHello {
+		if _, ok := conn.(*net.TCPConn); !ok {
+			return conn
+		}
+		tlsFragConn, err := tlsfrag.WrapConnFunc(conn.(*net.TCPConn), func(record []byte) (n int) {
+			return index
+		})
+		if err != nil {
```
Comment (hwh33) [inline chained/tlsfrag.go:22]: I feel like we should log here so that we have some indication that fragmentation was not used.

File: chained/https_impl.go [hunk 6]
```diff
@@ -82,6 +85,8 @@ func (impl *httpsImpl) dialServer(op *ops.Op, ctx context.Context) (net.Conn, er
 			if impl.tlsClientHelloSplitting {
 				tcpConn = hellosplitter.Wrap(tcpConn, splitClientHello)
 			}
+			tcpConn = tlsFragConn(tcpConn, impl.proxyConfig)
```
Comment (hwh33) [inline chained/https_impl.go]: This makes it look like we always fragment.  I think it'd be more readable if the setting was checked explicitly here, like the other settings.  Not a huge deal though.

File: chained/tlsmasq_impl.go [hunk 7]
```diff
@@ -142,6 +148,7 @@ func (impl *tlsMasqImpl) dialServer(op *ops.Op, ctx context.Context) (net.Conn,
 	if impl.tlsClientHelloSplitting {
 		tcpConn = hellosplitter.Wrap(tcpConn, splitClientHello)
 	}
+	tcpConn = tlsFragConn(tcpConn, impl.proxyConfig)
```
Comment (hwh33) [inline chained/tlsmasq_impl.go]: Same comment as on the HTTPS transport:
https://github.com/getlantern/flashlight/pull/1347#discussion_r1456459403

File: chained/tlsfrag.go [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+	"github.com/Jigsaw-Code/outline-sdk/transport/tlsfrag"
+	"github.com/getlantern/common/config"
+)
+
+func tlsFragConn(conn net.Conn, pc *config.ProxyConfig) net.Conn {
+	splitHello, index := splitHelloInfo(pc)
+	if splitHello {
+		if _, ok := conn.(*net.TCPConn); !ok {
+			return conn
+		}
```
Comment (hwh33) [inline chained/tlsfrag.go]: Wait does this work in a production context?  I thought we wrap the transport with our own type.

File: chained/tlsfrag.go (hunk 4 above)
Comment (myleshorton) [reply chained/tlsfrag.go]: I feel like you have a clearer picture as to what that should look like than I do. You want to just run with that and implement it on top of this PR?

File: chained/https_impl.go (hunk 6 above)
Comment (myleshorton) [reply chained/https_impl.go]: Yeah it's almost more like a filter, but tbh I don't love the duplicate logic of the other approach

File: chained/tlsfrag.go (hunk 4 above)
Comment (hwh33) [reply chained/tlsfrag.go]: No problem!

859f05b0cd6a0d11ae48bd1d4a554bbf65317dfc

File: chained/tlsfrag.go [hunk 9]
```diff
@@ -0,0 +1,53 @@
...
+	return conn
+}
+
+func makeFragFunc(pc *config.ProxyConfig) (tlsfrag.FragFunc, bool) {
+	fragStr, ok := pc.PluggableTransportSettings["tlsfrag"]
+	if !ok {
+		return nil, false
+	}
+
+	// fragStr should be of the form <frag func> or <frag func>:<func config>.
```
Comment (hwh33) [inline chained/tlsfrag.go:36]: Does this comment make sense to you @myleshorton?  I wasn't sure what would be the most straight-forward in terms of nomenclature or formatting.

Comment (hwh33) [reply chained/tlsfrag.go:36]: I added a line below this with an example

File: chained/tlsfrag.go (hunk 8 above)
Comment (myleshorton) [reply chained/tlsfrag.go]: Oh very good point....I wonder what it actually is there...will look.

File: chained/tlsfrag.go [hunk 10]
```diff
@@ -75,3 +73,17 @@ func makeFragFunc(pc *config.ProxyConfig) (tlsfrag.FragFunc, bool) {
...
+
+func (sc *streamConn) CloseRead() error {
+	return sc.Close()
+}
+
+func (sc *streamConn) CloseWrite() error {
+	return sc.Close()
+}
+
+var _ transport.StreamConn = (*streamConn)(nil)
```
Comment (hwh33) [inline chained/tlsfrag.go:117]: Nice, I always forget about this trick.

File: chained/tlsfrag_test.go [hunk 11]
```diff
@@ -20,7 +20,7 @@ func TestTLSFragConn(t *testing.T) {
 	resultType := reflect.TypeOf(result)
 	require.Equal(t, "*net.TCPConn", resultType.String())
 
-	pc.PluggableTransportSettings["tlsfrag_split_index"] = "1"
+	pc.PluggableTransportSettings["tlsfrag"] = "index:1"
```
Comment (hwh33) [inline chained/tlsfrag_test.go:35]: Ah, thank you.  Sorry about that!

File: chained/tlsfrag.go [hunk 12]
```diff
@@ -32,20 +34,41 @@ func makeFragFunc(pc *config.ProxyConfig) (tlsfrag.FragFunc, bool) {
...
 		}
 		return func(_ []byte) int { return index }, true
+	case "regex":
+		regex, err := regexp.Compile(cfg)
+		if err != nil {
+			log.Errorf("tlsfrag: bad regex specifier: %v", err)
+			return nil, false
+		}
+		return func(b []byte) int {
+			loc := regex.FindIndex(b)
```
Comment (hwh33) [inline chained/tlsfrag.go]: I'm not totally sure that this works.  It sounds like the `regexp` package expects the input to be UTF-8 code-points.  I'm not sure how that would behave with a ClientHello.  I think the text, like the value of the SNI, might be UTF-8 encoded.  But in theory there could be unexpected collisions with the binary data.  I guess that's unlikely?  But technically the behavior would be undefined... or just sort of non-intuitive.  It'd also be difficult to look for stuff like the SNI extension without knowing the SNI ahead of time.  You could look for the expected UTF-8 code-points, but that's pretty awkward IMO.

It's a bummer because specifying a split function as regex would be super powerful.  I thought it was an awesome idea.

Comment (hwh33) [reply chained/tlsfrag.go]: Oh, what if the regular expression expects to see the ClientHello as a hex-encoded string?  To be concrete, here you would have:
```
return func(b []byte) int {
    loc := regex.FindStringIndex(hex.EncodeToString(b))
    // handle loc == nil
    return loc[0]
}
```

File: chained/tlsfrag.go [hunk 13]
```diff
@@ -32,20 +34,41 @@ func makeFragFunc(pc *config.ProxyConfig) (tlsfrag.FragFunc, bool) {
...
+			if loc == nil {
+				return 0
+			}
+			return loc[0]
+		}, true
+	case "indexOf":
+		return func(b []byte) int {
+			index := bytes.Index(b, []byte(cfg))
```
Comment (hwh33) [inline chained/tlsfrag.go]: This has the same problem as the regular expression function, but could also be solved by treating the input as a hex-encoded string.

File: chained/https_impl.go [hunk 14]
```diff
@@ -85,7 +88,9 @@ func (impl *httpsImpl) dialServer(op *ops.Op, ctx context.Context) (net.Conn, er
 			if impl.tlsClientHelloSplitting {
 				tcpConn = hellosplitter.Wrap(tcpConn, splitClientHello)
 			}
-			tcpConn = tlsFragConn(tcpConn, impl.proxyConfig)
+			for _, wrapper := range impl.connWrappers {
+				tcpConn = wrapper(tcpConn, impl.proxyConfig)
+			}
```
Comment (hwh33) [inline chained/https_impl.go:93]: Oh this is nice!

**Comment by myleshorton** [comment]
OK I think this is all set @hwh33. I added more tests and a generic net.Conn wrapper func

File: chained/tlsfrag.go (hunk 12 above)
Comment (myleshorton) [reply chained/tlsfrag.go]: Interesting yeah I like the hex encoding idea, although I think it would throw the index off? Maybe we could figure out what the index should be though?

At least in that test with the test hello, the straight regex works, but I agree the regex theoretically is looking for UTF-8. I thought it might bork at non-UTF-8 bytes, but that doesn't appear to be true in the test at least.

File: chained/tlsfrag.go (hunk 13 above)
Comment (myleshorton) [reply chained/tlsfrag.go]: OK added the hex encoding with just dividing by two to get the index -- good idea!

Comment (myleshorton) [reply chained/tlsfrag.go]: Oh I ditched the indexOf btw because you can just use regex with a plain string to achieve the same thing (i.e. bytes.Index is the same as regex.FindIndex with a naked string.

**Comment by myleshorton** [comment]
Pulling this in to get a desktop version out. Thanks for the review @hwh33!
